### ISSUE TITLE
[no-inferrable-types] conflict in function params of any with default value

### ISSUE THREAD

**Issue by Mister-Hope** [typescript-eslint/typescript-eslint#3145]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```yaml
root: true

env:
  node: true
  es6: true

extends:
  - eslint:recommended
  - plugin:prettier/recommended

overrides:

  - files:
      - "*.ts"

    extends:
      - plugin:@typescript-eslint/eslint-recommended
      - plugin:@typescript-eslint/recommended
      - plugin:@typescript-eslint/recommended-requiring-type-checking
      - plugin:prettier/recommended

    plugins:
      - "@typescript-eslint"

    parserOptions:
      parser: "@typescript-eslint/parser"
      project:
        - ./tsconfig.json
        - ./tests/e2e/tsconfig.json

    rules:
      "@typescript-eslint/explicit-function-return-type":
        - warn
        - allowTypedFunctionExpressions: true

      "@typescript-eslint/naming-convention":
        - warn
        - selector: default
          format:
            - camelCase
          leadingUnderscore: allow
          trailingUnderscore: allow

        - selector:
            - variable
          format:
            - camelCase
            - UPPER_CASE
            - PascalCase
          leadingUnderscore: allow
          trailingUnderscore: allow

        - selector:
            - property
          format:
            - camelCase
            - PascalCase
          leadingUnderscore: allow
          trailingUnderscore: allow

        - selector: typeLike
          format:
            - PascalCase

      "@typescript-eslint/no-explicit-any":
        - warn
        - ignoreRestArgs: true

      "@typescript-eslint/no-unsafe-member-access":
        - warn

```

```TS
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type Mutation = (id: string, payload?: any) => void;

export const mutation: Mutation = (id, isLogin: boolean = true) => {
  id = isLogin.toString();
};

```

TsConfig:

```json
{
  "compilerOptions": {
    "allowJs": false,
    "allowSyntheticDefaultImports": true,
    "allowUnreachableCode": false,
    "allowUnusedLabels": false,
    "alwaysStrict": true,
    "baseUrl": ".",
    "declaration": true,
    "diagnostics": false,
    "downlevelIteration": true,
    "experimentalDecorators": true,
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "importHelpers": true,
    "inlineSourceMap": false,
    "inlineSources": false,
    "jsx": "preserve",
    "listEmittedFiles": true,
    "module": "esnext",
    "moduleResolution": "node",
    "noEmitHelpers": true,
    "noFallthroughCasesInSwitch": true,
    "noImplicitAny": true,
    "noImplicitReturns": true,
    "noImplicitThis": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "pretty": true,
    "resolveJsonModule": true,
    "removeComments": true,
    "sourceMap": true,
    "strict": true,
    "strictNullChecks": true,
    "strictPropertyInitialization": true,
    "target": "esnext",
    "paths": {
      "@/*": ["src/*"]
    }
  },
  "include": [
    "src/**/*.ts",
    "src/**/*.tsx",
    "src/**/*.vue",
    "tests/unit/**/*.ts",
    "typings/**/*.ts"
  ]
}
```

test/e2e/tsconfig.json:

```json
{
  "compilerOptions": {
    "target": "es5",
    "lib": ["es5", "dom"],
    "types": ["cypress"]
  },
  "include": ["**/*.ts"]
}
```

![image](https://user-images.githubusercontent.com/33315834/109962417-1df8b580-7d26-11eb-8da5-0068d8a99493.png)
![image](https://user-images.githubusercontent.com/33315834/109962457-2b15a480-7d26-11eb-8fdf-4f92f03a5ca0.png)

I am expecting `@typescript-eslint/no-inferrable-types` not throwing any errors.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.16.1` |
| `@typescript-eslint/parser`        | `4.16.1` |
| `TypeScript`                       | `4.2.2` |
| `ESLint`                           | `7.21.0` |
| `node`                             | `14.15.5` |

**Comment by bradzacher** [comment]
`no-inferrable-types` does not usw type information and only works on the AST. This means it does not and cannot know the actual type of the variable.

There's nothing we can do about this - you'll need to use a disable comment.

**Comment by Mister-Hope** [comment]
Can't we build an enhance version of it in type checking? Then we can disable it and enable the enhance ones.

I am running in to this situation, where the types are provided bt packages, while I am going to use a lot og default param

**Comment by bradzacher** [comment]
Consider up steaming improvements to the libraries to remove `any` from their API. For example here you could replace the any with a generic for stricter types.

There's very little call for a type-aware version of the inferable types rule. In addition we don't want to have duplicate versions of the rule in this plugin - it just makes life harder for people to figure out how to configure things.

Finally, won't be adding type information to the base rule because that is a large breaking change which restricts a large portion of users from using it (all users that don't use type information).

If your project has a hard requirement on a type-aware version of this rule - our code is MIT licenced, so you're free to fork it and change it to suit your projects needs.
